subworkflows/local/preprocessing.nf

```diff
@@ -0,0 +1,83 @@
+nextflow.enable.dsl = 2
+
+//includes
+include { FORCE_GENERATE_TILE_ALLOW_LIST }         from '../../modules/local/force-generate_tile_allow_list'
+include { FORCE_GENERATE_ANALYSIS_MASK }           from '../../modules/local/force-generate_analysis_mask'
+include { PREPROCESS_CONFIG }                      from '../../modules/local/preprocess_force_config'
+include { FORCE_PREPROCESS }                       from '../../modules/local/force-preprocess'
+include { MERGE as MERGE_BOA; MERGE as MERGE_QAI } from '../../modules/local/merge'
+
+//Closure to extract the parent directory of a file
+def extractDirectory = { it.parent.toString().substring(it.parent.toString().lastIndexOf('/') + 1 ) }
+
+workflow PREPROCESSING {
+
+    take:
+        data
+        dem
+        wvdb
+        cube_file
+        aoi_file
+
+    main:
+
+        ch_versions = Channel.empty()
+
+        FORCE_GENERATE_TILE_ALLOW_LIST( aoi_file, cube_file )
+        ch_versions = ch_versions.mix(FORCE_GENERATE_TILE_ALLOW_LIST.out.versions)
+
+        FORCE_GENERATE_ANALYSIS_MASK( aoi_file, cube_file )
+        ch_versions = ch_versions.mix(FORCE_GENERATE_ANALYSIS_MASK.out.versions)
+
+        //Group masks by tile
+        masks = FORCE_GENERATE_ANALYSIS_MASK.out.masks.flatten().map{ x -> [ extractDirectory(x), x ] }
+
+        // Preprocessing configuration
+        PREPROCESS_CONFIG( data, cube_file, FORCE_GENERATE_TILE_ALLOW_LIST.out.tile_allow, dem, wvdb )
+        ch_versions = ch_versions.mix(PREPROCESS_CONFIG.out.versions.first())
+
+        // Main preprocessing
+        FORCE_PREPROCESS( PREPROCESS_CONFIG.out.preprocess_config_and_data)
+        ch_versions = ch_versions.mix(FORCE_PREPROCESS.out.versions.first())
+
+        //Group by tile, date and sensor
+        boa_tiles = FORCE_PREPROCESS.out.boa_tiles.flatten().map{ [ "${extractDirectory(it)}_${it.simpleName}", it ] }.groupTuple()
+        qai_tiles = FORCE_PREPROCESS.out.qai_tiles.flatten().map{ [ "${extractDirectory(it)}_${it.simpleName}", it ] }.groupTuple()
+
+        //Find tiles to merge
+        boa_tiles_to_merge = boa_tiles.filter{ x -> x[1].size() > 1 }
+                                .map{ [ it[0].substring( 0, 11 ), it[1] ] }
```

Comment:
This is an example as to why we generally don't like to encode things in file/directory names, off by one errors in directory paths etc., can occur and the whole thing falls over.